izpack-compiler/src/main/java/com/izforge/izpack/compiler/packager/impl/PackagerBase.java

```diff
@@ -381,6 +384,41 @@ protected boolean packSeparateJars()
         return info != null && info.getWebDirURL() != null;
     }
 
+    private List<DynamicVariable> buildVariableList()
+    {
+        DirectedGraph<DynamicVariable> graph = new DirectedGraph<DynamicVariable>();
+        DynamicVariable rootVar = new DynamicVariableImpl("", "");
+
+        for (List<DynamicVariable> dynVariables : dynamicVariables.values())
+        {
+            for (DynamicVariable var : dynVariables)
+            {
+                graph.addEdge(rootVar, var);
+                for (String childName : var.getUnresolvedVariableNames())
+                {
+                    List<DynamicVariable> childVars = dynamicVariables.get(childName);
+                    if (childVars != null)
+                    {
+                        for (DynamicVariable childVar : childVars)
+                        {
+                            graph.addEdge(var, childVar);
+                        }
+                    }
+                }
+            }
+        }
+
+        Iterator<DynamicVariable> it = new BreadthFirstIterator<DynamicVariable>(graph, rootVar);
+        List<DynamicVariable> nodes = new ArrayList<DynamicVariable>();
+        while (it.hasNext()) {
+            nodes.add(it.next());
+        }
+        nodes.remove(rootVar);
```

Comment:
Using the breadth first search will not give us the intented result! A BFS will give us the shortest way to a certain element, but what we need is the longest way (without loops)!

The following example will demonstrate this:
```
<dynamicvariables>
  <variable name="var1" value="${var2} - ${var4}" />
  <variable name="var2" value="${var3}" />
  <variable name="var3" value="${var4}" />
  <variable name="var4" value="value" />
</dynamicvariables>
```
In this example var4 is a successor of both va1 and var3. A BFS would deliver a sequence like 
``var1 - var2 - var4 - var3 - var4``. But in a BFS only the first occurence of var4 will be recorded. So after reversing the nodes-list the var3 will be evaluated before var4, which is not intended.

A second problem in the implementation above is, that all variables are a successor of the root element. So in fact the whole dependencies collaps to the simple dependency from the root element.